src/ucs/debug/memtrack.c

```diff
@@ -214,6 +214,34 @@ int ucs_posix_memalign(void **ptr, size_t boundary, size_t size, const char *nam
     return ret;
 }
 
+int ucs_posix_memalign_realloc(void **ptr, size_t boundary, size_t size,
+                               const char *name)
+{
+    void *tmp;
+    int ret;
+
+    /* first try to realloc() - the region may be extended (not guaranteed) */
+    tmp = ucs_realloc(*ptr, size, name);
+    if (!tmp) {
+        return -1;
+    }
+    if ((uintptr_t)tmp % boundary == 0) {
+        *ptr = tmp;
+        return 0;
+    }
+
+#if HAVE_POSIX_MEMALIGN
+    ret = posix_memalign(ptr, boundary, size);
+#else
+#error "Port me too"
+#endif
+    if (ret == 0) {
+        memcpy(*ptr, tmp, size);
```

Comment:
what is size is smaller than previous size? (+add test for it)

Reply:
do you think it's worth going through the hash-table to find the old size of the allocation?

Reply:
Also, the last call to `ucs_posix_memalign_realloc()` is a "best-effort" test for this... but I don't see a good way to ensure the space will indeed be realloc-ed (and not just shrinked in-place).

Reply:
if the new size is smaller then it will be segfault during memcpy()
so must handle it somehow
try https://man7.org/linux/man-pages/man3/malloc_usable_size.3.html

Reply:
Maybe I'm missing something, but if the `realloc()` was successful - the worst I can see happening is we copy some garbage (since the buffer is now with the new size, just not aligned).

Reply:
right, since size is the new size. but if new size is larger, how much we will copy from the old small buffer?

Reply:
still don't see it... after the call to `ucs_realloc()` succeeds - there is only one buffer (`tmp` points to it) and it's size is the new side (`size`). If the old size was smaller - the difference can potentially be garbage data (but allocated accessible).

Reply:
ok. so it's just optimization to not copy excess data. can we use malloc_usabale_size to copy only what is needed?

Reply:
sure, I suppose we can. It's not fast-path anyway...